grpc/src/main/java/com/linecorp/armeria/common/grpc/GrpcJsonMarshallerBuilder.java

```diff
@@ -61,13 +64,27 @@ public GrpcJsonMarshallerBuilder jsonMarshallerCustomizer(
         return this;
     }
 
+    /**
+     * Specifies the {@link GrpcJsonMarshallType} that should be created.
+     */
+    public GrpcJsonMarshallerBuilder grpcJsonMarshallType(GrpcJsonMarshallType grpcJsonMarshallType) {
+        this.grpcJsonMarshallType = requireNonNull(grpcJsonMarshallType);
+        return this;
+    }
+
     /**
      * Returns a newly-created {@link GrpcJsonMarshaller} with the specified {@link ServiceDescriptor}.
      */
     public GrpcJsonMarshaller build(ServiceDescriptor serviceDescriptor) {
         requireNonNull(serviceDescriptor, "serviceDescriptor");
-        return new DefaultJsonMarshaller(
-                GrpcJsonUtil.jsonMarshaller(ImmutableList.copyOf(serviceDescriptor.getMethods()),
-                                            jsonMarshallerCustomizer));
+        if (grpcJsonMarshallType == GrpcJsonMarshallType.DEFAULT) {
+            return new DefaultJsonMarshaller(serviceDescriptor, jsonMarshallerCustomizer);
+        } else if (grpcJsonMarshallType == GrpcJsonMarshallType.PROTOBUF_JACKSON) {
+            return GrpcJsonUtil.protobufJacksonJsonMarshaller(serviceDescriptor, jsonMarshallerCustomizer);
+        } else if (grpcJsonMarshallType == GrpcJsonMarshallType.UPSTREAM) {
+            return UpstreamJsonMarshaller.INSTANCE;
```

Comment:
`GrpcJsonMarshallType.UPSTREAM` and `jsonMarshallerCustomizer` are mutually exclusive. 
Originally, this builder is designed for `protobuf-jackson`. 
How about adding some factory methods instead of `GrpcJsonMarshallType`?
```java
// Choose the JSON codec depending on a proto syntax.
GrpcJsonMarshaller.of();

// Always uses protobuf-jackon
GrpcJsonMarshaller.ofJackson(...);
GrpcJsonMarshaller.builderForJackson()
                  .jsonMarshallerCustomizer(...)
                  .build(ServiceDescriptor);

// Always uses protobuf-util's JsonFormat
GrpcJsonMarshaller.ofGson(...);
```

Reply:
Good point.

I've updated my thoughts on the default implementation in the PR description.

I'm thinking it might be better to leave the default as-is, and just providing a `gson` implementation which users can `opt-in`. This can be slightly inconvenient since users need to add compatibility for both client and server-side, but at least there will be no breaking changes.

```
GrpcService.builder()
           .jsonMarshallerFactory(sd -> GrpcJsonMarshaller.ofGson())
           .build();
...
GrpcClients.builder(server.httpUri(GrpcSerializationFormats.JSON))
           .jsonMarshallerFactory(sd -> GrpcJsonMarshaller.ofGson())
           .build(Proto2ServiceBlockingStub.class);
```

Reply:
> I'm thinking it might be better to leave the default as-is

Sounds good. 👍